### ISSUE TITLE
ImageVisual doesn't support masked array

### ISSUE THREAD

**Issue by EtienneCmb** [vispy/vispy#1509]
One nice feature when plotting images using Matplotlib is that it support [masked array](https://docs.scipy.org/doc/numpy-1.13.0/reference/maskedarray.html) which mean that array values with a `True` set to the mask are not displayed (and not take into account in the colormap)

```py
import numpy as np

import matplotlib.pyplot as plt

from vispy import scene
from vispy.scene.cameras import PanZoomCamera

n = 1000

data = np.ones((n, n))
y, x = np.ogrid[-n / 2:n / 2, -n / 2:n / 2]
mask = x**2 + y**2 >= (n / 2)**2
data[mask] = 0.
image = np.ma.masked_array(data, mask=mask)

# Matplotlib :
plt.imshow(image)
plt.show()

# Vispy :
cam = PanZoomCamera(rect=(0, 0, n, n))
sc = scene.SceneCanvas(bgcolor='black', show=True)
wc = sc.central_widget.add_view(camera=cam)
im_vispy = scene.visuals.Image(image, parent=wc.scene)
sc.app.run()

```

![image](https://user-images.githubusercontent.com/15892073/44401380-78411a80-a51c-11e8-8b38-5c0303a3b21f.png)

**Comment by djhoese** [comment]
Good idea. First question I have is what happens if you fill the masked area with NaNs? Actually for the ImageVisual that might be unspecified behavior since data is converted to 8-bit integers.

I hesitate to add this functionality to the Visual classes since they are one level above the low-level gloo interface. This is just a hesitation, not a strong feeling. Also if we add the support for masked arrays for one visual then some users might expect it for all visuals. I could see ImageVisual and LineVisual using this with little effort. It may be a lot more difficult to support it for other visuals and some it may not make sense at all.

**Comment by EtienneCmb** [comment]
I've tested the following code :
```py
import numpy as np

import matplotlib.pyplot as plt

from vispy import scene
from vispy.scene.cameras import PanZoomCamera

n = 10

data = np.ones((n, n))
y, x = np.ogrid[-n / 2:n / 2, -n / 2:n / 2]
mask = x**2 + y**2 >= (n / 2)**2
data[~mask] = np.nan

# Vispy :
cam = PanZoomCamera(rect=(0, 0, n, n))
sc = scene.SceneCanvas(bgcolor='black', show=True)
wc = sc.central_widget.add_view(camera=cam)
im_vispy = scene.visuals.Image(data, parent=wc.scene)
sc.app.run()
```
Producing :
![image](https://user-images.githubusercontent.com/15892073/44404340-ac203e00-a524-11e8-9e5e-ecd4a3ac6fdd.png)

Tested with `np.nan` and `np.inf`. I think vertex buffers take `np.float32` arrays no?

I don't think every visuals should handle masked array. For example, if I remember correctly, `LineVisual` provide a `connect` input and you can already specify which point to connect.

**Comment by djhoese** [comment]
So it sounds like the ImageVisual is the visual that lacking this functionality and would benefit the most from it. It may not be the best solution but we could specifically check for NaN values (or an equivalent) in the fragment shader and discard the pixels: https://www.opengl.org/sdk/docs/tutorials/ClockworkCoders/discard.php

This may not work for the current implementation of the ImageVisual since it is creating an 8-bit texture. Hhhhmmm, although if it is an RGBA texture we could overwrite the alpha channel to 0.

> I think vertex buffers take np.float32 arrays no?

I think so.
